integration_test/tests/prepared_statement_tests.rs

```diff
@@ -0,0 +1,186 @@
+            // Calling SQLExecute before SQLPrepare is invalid.
+            assert_eq!(SqlReturn::ERROR, SQLExecute(stmt));
+
+            dbg!(get_sql_diagnostics(HandleType::Stmt, stmt as Handle));
+        }
+        disconnect_and_close_handles(dbc, stmt);
+    }
+
+    #[test]
+    fn test_prepare_get_resultset_metadata() {
+        let (dbc, stmt) = connect_and_allocate_statement();
+
+        unsafe {
+            let mut query: Vec<WideChar> = cstr::to_widechar_vec("select * from example");
+            query.push(0);
+            // Preparing a statement.
+            // Only the result set metadata are retrieved and stored
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLPrepareW(stmt as HStmt, query.as_ptr(), NTS as SmallInt as i32),
+                "{}",
+                get_sql_diagnostics(HandleType::Stmt, stmt as Handle)
+            );
+
+            // Retrieve result set metadata
+            get_column_attributes(stmt as Handle, 2);
+
+            disconnect_and_close_handles(dbc, stmt);
+        }
+    }
+
+    #[test]
+    fn test_error_fetch_before_execute() {
+        let (dbc, stmt) = connect_and_allocate_statement();
+
+        unsafe {
+            let mut query: Vec<WideChar> = cstr::to_widechar_vec("select * from example");
+            query.push(0);
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLPrepareW(stmt as HStmt, query.as_ptr(), NTS as SmallInt as i32),
+                "{}",
+                get_sql_diagnostics(HandleType::Stmt, stmt as Handle)
+            );
+
+            // Retrieve result set metadata
+            get_column_attributes(stmt as Handle, 2);
+
+            assert_eq!(SqlReturn::ERROR, SQLFetch(stmt as HStmt),);
+
+            dbg!(get_sql_diagnostics(HandleType::Stmt, stmt as Handle));
+
+            disconnect_and_close_handles(dbc, stmt);
+        }
+    }
+
+    #[test]
+    fn test_prepare_execute_retrieve_data() {
+        let (dbc, stmt) = connect_and_allocate_statement();
+
+        unsafe {
+            let mut query: Vec<WideChar> = cstr::to_widechar_vec("select * from example");
+            query.push(0);
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLPrepareW(stmt as HStmt, query.as_ptr(), NTS as SmallInt as i32),
+                "{}",
+                get_sql_diagnostics(HandleType::Stmt, stmt as Handle)
+            );
+
+            // Retrieve result set metadata
+            get_column_attributes(stmt as Handle, 2);
+
+            // Executing the prepared statement.
+            // The $sql pipeline is now executed and the result set cursor.
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLExecute(stmt as HStmt),
+                "{}",
+                get_sql_diagnostics(HandleType::Stmt, stmt as Handle)
+            );
+
+            // A prepared statement can be executed multiple times.
+            assert_eq!(
```

Comment:
is the main difference between this test and the following that this calls SQLExecute twice before fetching?

Reply:
I was wondering the same, it seems redundant since this is done in `test_prepare_execute_multiple_times`

Reply:
Yes, this is an oversight.
I split it and the last SQLExecute should not be there.